torch/csrc/jit/tensorexpr/kernel.cpp

```diff
@@ -1554,6 +1563,64 @@ Tensor* TensorExprKernel::computeSum(const torch::jit::Value* v) {
       reduction_info.reductionDims);
 }
 
+Tensor* TensorExprKernel::computeSoftmax(const torch::jit::Value* v) {
+  // Softmax is computed as follows:
+  //    softmax(vi) = exp(vi) / sum(exp(vi))
+  //
+  // This is implemented as 3 loopnests:
+  //   - First loop computes exp for every element in v.
+  //   - Second loop computes the sum over the softmax dim.
+  //   - Final loop computes softmax for every element in v.
+
+  TORCH_INTERNAL_ASSERT(v->node()->inputs().size() == 3);
+  auto output_dims = dimsFromSizes(sizesForValue(v));
+
+  // We do not handle None for dims (input 1) because that is supposed to
+  // be deprecated.
+  TORCH_INTERNAL_ASSERT(v->node()->input(1)->node()->kind() == prim::Constant);
+  size_t softmax_dim = v->node()->input(1)->node()->i(attr::value);
+
+  // The sum needed for Softmax should be calculated along the softmax
+  // dimension, which will be dimension to reduce.
+  std::vector<DimArg> dim_args;
+  for (size_t i = 0; i < output_dims.size(); ++i) {
+    if (i != softmax_dim) {
+      dim_args.push_back(output_dims[i]);
+    }
+  }
+
+  auto e = computeOneOperand(
+      "aten_softmax_exp", v, [](const ExprHandle& a) { return exp(a); });
+  auto sum = Reduce(
+      "aten_softmax_sum",
+      dim_args,
+      Sum(),
+      [&](ParameterList& indices) {
+        // Reduce will have the softmax dimension as the inner most loop.
+        // So, the innermost index in the incoming indices will refer to the
+        // softmax dimension. Update the incoming indices by moving that
+        // softmax dimension index to the appropriate position.
+        auto new_indices = indices;
+        for (size_t i = softmax_dim; i < indices.size() - 1; ++i) {
+          new_indices[i + 1] = indices[i];
+        }
+        new_indices[softmax_dim] = indices[indices.size() - 1];
+        return e->call(new_indices);
+      },
+      {output_dims[softmax_dim]});
+  return Compute("aten_softmax", output_dims, [&](ParameterList& indices) {
+    // The sum tensor uses all indices other than the softmax dimension
+    // index.
+    std::vector<ExprHandle> new_indices;
+    for (size_t i = 0; i < indices.size(); ++i) {
+      if (i != softmax_dim) {
+        new_indices.push_back(indices[i]);
+      }
+    }
+    return e->call(indices) / sum->call(new_indices);
```

Comment:
The default lowering should inline e->call(...). Either call an inline directly, or just manually. Either way is fine with me.

Reply:
Are you suggesting I inline `e->call(...)` here? Why is that necessary?

Reply:
We can't call inline without constructing a Loopnest (which we don't want to do here) - the only choice we have is to inline manually.

Reply:
As discussed offline, this is currently not possible with the infrastructure we have. But I am planning to fix inlining to handle these cases, in a separate PR.